File: python/difference-of-squares/difference_of_squares.py
def difference_of_squares(N: int) -> int:
    """Return the the difference of squares for numbers up to N.

    :param N: int - the Nth natural number.
    :return: int - the difference of the square of the sum and the sum of the squares.
    """
    return square_of_sum(N) - sum_of_squares(N)


def square_of_sum(N: int) -> int:
    """Returns the square of the sum of the first N natural numbers."""
    return sum(range(N + 1)) ** 2


def sum_of_squares(N: int) -> int:
    """Returns the sum of the squares of the first N natural numbers."""
    return sum(i * i for i in range(N + 1))

File: ruby/difference-of-squares/difference_of_squares.rb
class Squares
  attr_reader :number, :natural_numbers

  def initialize(number)
    @number = number
    @natural_numbers = (0..number)
  end

  def difference
    square_of_sum - sum_of_squares
  end

  def square_of_sum
    sum ** 2
  end

  def sum_of_squares
    squares.inject { |sum, i| sum + i }
  end

  private

  def sum
    natural_numbers.inject { |sum, i| sum + i }
  end

  def squares
    natural_numbers.map { |i| i * i }
  end
end

module BookKeeping
  VERSION = 3
end
